crates/codegen/language/definition/src/model/utils/mod.rs

```diff
@@ -1,5 +1,78 @@
+use crate::model::{self, Item};
+use semver::Version;
+use std::collections::BTreeSet;
+
 mod identifier;
 mod version_specifier;
 
 pub use identifier::*;
 pub use version_specifier::*;
+
+/// Collects all versions that changes the language grammar in a breaking way.
+pub fn collect_breaking_versions(lang: &model::Language) -> BTreeSet<Version> {
+    let first = lang.versions.first().cloned().unwrap();
+    let mut res = BTreeSet::from_iter([first]);
+
+    fn add_spec(res: &mut BTreeSet<Version>, spec: &Option<model::VersionSpecifier>) {
+        let Some(spec) = spec else {
+            return;
+        };
+
+        match spec.clone() {
```

Comment:
Thoughts on using `.clone()` on specific versions instead of cloning the throw away `VersionSpecifier`?

Reply:
Ideally, the function argument should be owned if the argument is meant to be consumed (rather than accepting a `&` and cloning inside).

However, since we return early on `None` I decided to accept a ref and only clone internally if needed. This is a closure (and is inherently coupled to the code surrounding it) I decided to just leave it like that, since it strikes a good balance between readability (not having to `.clone()` in every call below) and trying to do reasonable thing perf-wise (not always cloning, as `None`s are abundant and the enum size is not that small).

Now, about specifically `clone`ing in the match: in general, moves are very cheap since it's semantically a bitwise copy; moreover, since the data is owned, this is easily optimizable since the data is guaranteed to not be aliased (thanks to Rust rules), so unless the logic is very hot, there shouldn't be a noticeable difference in cloning the sub-fields vs cloning the trivial container and moving the sub-fields, instead (and `Never` variants are, I think, quite a bit rarer than `None` so I decided to not optimize for that case).

Since the perf difference should be nil (or minimal, at worst), I decided to clone once at the top to not repeat `.clone()` multiple times in the branches below for clarity.

That was my thinking when I tried to strike a good balance but I'm open; let me know if you'd prefer to call `.clone()` in the branches!